File: 0557-reverse-words-in-a-string-iii/0557-reverse-words-in-a-string-iii.java
public class Solution {
    public String reverseWords(String s) {
        char[] ch = s.toCharArray();
        for(int i=0;i<ch.length;i++){
            //if the character is not a space
            if(ch[i]!=' '){
                int j = i;
                //move j to the end of the word till you encounter a space
                while(j + 1<ch.length && ch[j+1]!=' '){
                    j++;
                }
                rev(ch,i,j);
                i = j;
            }
        }
        return new String(ch);
    }
    private void rev(char[] ca, int i, int j) {
        for (; i < j; i++, j--) {
            char tmp = ca[i];
            ca[i] = ca[j];
            ca[j] = tmp;
        }
    }
}
  

File: 0557-reverse-words-in-a-string-iii/0557-reverse-words-in-a-string-iii.cpp
class Solution {
public:
    string reverseWords(string s) {
        
        istringstream iss(s); // Used to read the words from string 's'
        string word, result;

        // This loop will run until there are no more words to read
        while (iss >> word) {
            // reversing each word 
            reverse(word.begin(), word.end());
            // Storing the result with reversed words followed by a space
            result += word + " ";
        }

        // This is done to ensure that there is no extra space in the end of the result string
        if (!result.empty()) 
        result.pop_back();
        
        return result;
    }
};